PARCEL_2_RFC.md

```diff
@@ -0,0 +1,1285 @@
+of Parcel that your plugin supports:
+
+```json
+{
+  "name": "parcel-transform-imagemin"
+  "engines": {
+    "parcel": "2.x"
+  }
+}
+```
+
+If you do not specify this field, Parcel will output a warning:
+
+```
+Warning: The plugin "parcel-transform-typescript" needs to specify a `package.json#engines.parcel` field with the supported Parcel version range.
+```
+
+If you do specify the parcel engine field and the user is using an incompatible
+version of Parcel, they will see an error:
+
+```
+Error: The plugin "parcel-transform-typescript" is not compatible with the
+current version of Parcel. Requires "2.x" but the current version is "3.1.4"
+```
+
+Parcel uses node-semver to match version ranges.
+
+## Plugin APIs
+
+There are several different types of plugins. They all look very similar, but
+are kept separate so we can have strict contracts one what each one is allowed
+to do.
+
+There are some rules that should be followed across every type of plugin:
+
+- **Stateless** — Avoid any kind of state, it will likely be the source of bugs
+  for your users. For example, the same transform may exist in multiple
+  separate workers which are not allowed to communicate with one another, state
+  will not work as expected.
+- **Pure** — Given the same input, a plugin must produce the same output, and
+  you must not have any observable side effects, or implicit dependencies.
+  Otherwise Parcel's caching will break and your users will be sad. You should
+  never have to tell users to delete their caches.
+
+The plugin APIs all follow a common shape:
+
+```js
+import { NameOfPluginType } from '@parcel/plugin';
+
+export default new NameOfPluginType({
+  async methodName(opts: JSONObject): Promise<JSONObject> {
+    return result;
+  },
+});
+```
+
+They are made up of modules with well-known named exports of async functions
+that:
+
+- Accept a strictly validated JSON-serializable `opts` object.
+- Return a strictly validated JSON-serializable `vals` object.
+
+If something you need is not being passed through `opts`, please come talk to
+the Parcel team about it. Avoid trying to get information yourself from other
+sources, especially from the file system.
+
+### Resolvers
+
+Resolvers get called with an asset request (consisting of a source file path
+and the specifier of what is being requested) which it then attempts to
+resolve. If the resolver isn't sure how to handle a request, it can also return
+`null` and pass it to the next resolver in the chain.
+
+```js
+import { Resolver } from '@parcel/plugin';
+
+export default new Resolver({
+  async resolve({ assetRequest }) {
+    // ...
+    return { filePath } || null;
+  },
+});
+```
+
+### Transforms
+
+Transforms _transform_ single assets as they are discovered and added to the
+asset graph. They mostly call out to different compilers and preprocessors.
+
+```js
+import { Transform } from '@parcel/plugin';
+
+export default new Transform({
+  async config({ asset }) {
```

Comment:
would be nice to add js-doc to all methods below.
also, what is the _type_ of the asset passed ?

Reply:
The apis are not finalized yet